pkg/controllers/nodeoverlay/store.go

```diff
@@ -0,0 +1,287 @@
+}
+
+type InstanceTypeUpdate struct {
+	Price    map[string]*PriceUpdate
+	Capacity *CapacityUpdate
+}
+
+type InstanceTypeStore struct {
+	updates map[string]map[string]*InstanceTypeUpdate
+	mu      sync.RWMutex
+}
+
+func NewInstanceTypeStore() *InstanceTypeStore {
+	return &InstanceTypeStore{
+		updates: map[string]map[string]*InstanceTypeUpdate{},
+	}
+}
+
+func (s *InstanceTypeStore) UpdateStore(updatedStore map[string]map[string]*InstanceTypeUpdate) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.updates = map[string]map[string]*InstanceTypeUpdate{}
+	for nodePoolName, v := range updatedStore {
+		s.updates[nodePoolName] = map[string]*InstanceTypeUpdate{}
+		for overlayName, itUpdate := range v {
+			s.updates[nodePoolName][overlayName] = &InstanceTypeUpdate{
+				Price:    itUpdate.Price,
+				Capacity: itUpdate.Capacity,
+			}
+		}
+	}
+}
+
+func (s *InstanceTypeStore) FindInstanceTypesUpdate(nodePoolName string, it *cloudprovider.InstanceType) (*InstanceTypeUpdate, bool) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	instanceTypeList, foundNodePool := s.updates[nodePoolName]
+	if !foundNodePool {
+		return nil, false
+	}
+	instanceTypeUpdate, foundInstanceTypeUpdate := instanceTypeList[it.Name]
+	return instanceTypeUpdate, foundInstanceTypeUpdate
+}
+
+// updateInstanceTypeCapacity add a new Capacity overlay update to the associated instance type.
+// This add an Capacity update to the store once after it has been validated.
+func (s *InstanceTypeStore) updateInstanceTypeCapacity(nodePoolName string, instanceTypeName string, nodeOverlay v1alpha1.NodeOverlay) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if nodeOverlay.Spec.Capacity == nil {
+		return
+	}
+
+	_, foundNodePool := s.updates[nodePoolName]
+	if !foundNodePool {
+		s.updates[nodePoolName] = map[string]*InstanceTypeUpdate{}
+	}
+	_, foundInstanceTypeUpdate := s.updates[nodePoolName][instanceTypeName]
+	if !foundInstanceTypeUpdate {
+		s.updates[nodePoolName][instanceTypeName] = &InstanceTypeUpdate{Price: map[string]*PriceUpdate{}, Capacity: &CapacityUpdate{OverlayUpdate: corev1.ResourceList{}}}
+	}
+
+	if s.updates[nodePoolName][instanceTypeName].Capacity == nil {
+		s.updates[nodePoolName][instanceTypeName].Capacity = &CapacityUpdate{
+			OverlayUpdate:           nodeOverlay.Spec.Capacity,
+			lowestWeightOverlayName: nodeOverlay.Name,
+			lowestWeight:            nodeOverlay.Spec.Weight,
+		}
+	} else {
+		for resource, qun := range nodeOverlay.Spec.Capacity {
+			s.updates[nodePoolName][instanceTypeName].Capacity.OverlayUpdate[resource] = qun
+		}
+		s.updates[nodePoolName][instanceTypeName].Capacity.lowestWeight = nodeOverlay.Spec.Weight
+		s.updates[nodePoolName][instanceTypeName].Capacity.lowestWeightOverlayName = nodeOverlay.Name
+	}
+
+}
+
+func (s *InstanceTypeStore) isCapacityUpdateConflicting(nodePoolName string, instanceTypeName string, nodeOverlay v1alpha1.NodeOverlay) (string, bool) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	_, foundNodePool := s.updates[nodePoolName]
```

Comment:
nit: per standard practice, I'd just use `ok` everywhere for these checks. If it's unclear what `ok` means, I think that's an indicator that the name of the map or the key can be imrpoved. In this case though, I think those are fine and just switching to `ok` is good.